File: 801-900/0872-Leaf-Similar/0872.java
class Solution {
  public boolean leafSimilar(TreeNode root1, TreeNode root2) {
    List<Integer> seq1 = new ArrayList<Integer>();
    if (root1 != null) {
      dfs(root1, seq1);
    }

    List<Integer> seq2 = new ArrayList<Integer>();
    if (root2 != null) {
      dfs(root2, seq2);
    }

    return seq1.equals(seq2);
  }

  public void dfs(TreeNode node, List<Integer> seq) {
    if (node.left == null && node.right == null) {
      seq.add(node.val);
    } else {
      if (node.left != null) {
        dfs(node.left, seq);
      }
      if (node.right != null) {
        dfs(node.right, seq);
      }
    }
  }
}

File: 801-900/0872-Leaf-Similar/0872.py
class Solution:
  def leafSimilar(self, root1: TreeNode, root2: TreeNode) -> bool:
    def dfs(node: TreeNode):
      if not node.left and not node.right:
        yield node.val
      else:
        if node.left:
          yield from dfs(node.left)
        if node.right:
          yield from dfs(node.right)

    seq1 = list(dfs(root1)) if root1 else list()
    seq2 = list(dfs(root2)) if root2 else list()
    return seq1 == seq2